aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -16,25 +16,79 @@ namespace impl {
 
 // Assume T is decayed
 template <typename T>
-using not_ok_to_box = guts::negation<guts::disjunction<
+using ok_to_box = guts::disjunction<
     std::is_constructible<IValue, T>,
     // TensorOptions are not directly constructible into IValue,
     // but torch::jit::push knows how to handle them
     std::is_same<TensorOptions, T>,
     // void returns are ok
-    std::is_same<void, T>>>;
+    std::is_same<void, T>>;
 
-// TODO boxing should be ok for all kernels. Then remove not_ok_to_box and supports_boxing.
+// TODO boxing should be ok for all kernels. Then remove ok_to_box and supports_boxing.
 
-template <class Result, class... Args>
-using supports_boxing =
+template <typename Result>
+using supports_boxing_result =
   guts::negation<guts::disjunction<
     std::is_lvalue_reference<Result>,
-    not_ok_to_box<Result>,
-    std::is_same<IntArrayRef, Result>,
-    not_ok_to_box<std::decay_t<Args>>...
+    guts::negation<ok_to_box<Result>>,
+    std::is_same<IntArrayRef, Result>
   >>;
 
+template <class Result, class... Args>
+using supports_boxing =
+  guts::conjunction<
+    supports_boxing_result<Result>,
+    ok_to_box<std::decay_t<Args>>...
+  >;
+
+template <typename T, std::enable_if_t<!c10::impl::ok_to_box<T>::value>* = nullptr>
+inline bool pushIValueOrCannotBox(std::vector<c10::IValue>& stack, const T& v) {
+  torch::jit::push(stack, "cannot box");
+  return false;
+}
+template <typename T, std::enable_if_t<c10::impl::ok_to_box<T>::value>* = nullptr>
+inline bool pushIValueOrCannotBox(std::vector<c10::IValue>& stack, const T& v) {
+  torch::jit::push(stack, v);
+  return true;
+}
+
+// boxArgumentsIntoStack takes the arguments and pushes them as IValues onto the stack.
+// In case the argument cannot be converted to IValue, the function pushes "cannot box"
+// IValue string. Return value - whether all of the arguments could be converted to IValues
+inline bool boxArgumentsIntoStack(std::vector<c10::IValue>& stack) {
+  return true;
+}
+template<typename Item>
+inline bool boxArgumentsIntoStack(std::vector<c10::IValue>& stack, const Item& item) {
+  return pushIValueOrCannotBox(stack, item);
+}
+template<typename Item, typename... Rest>
+inline bool boxArgumentsIntoStack(std::vector<c10::IValue>& stack, const Item& item, Rest... other_items) {
+  auto res = pushIValueOrCannotBox(stack, item);
+  return boxArgumentsIntoStack(stack, other_items...) && res;
+}
+
+template<class Result>
```

Comment:
nit: those three variants could be combined into one with a `if_constexpr`, which would make it a bit easier to read in my opinion.